### DISCUSSION TITLE
How to disconnect model in many to many relation on delete

### DISCUSSION THREAD

**Discussion by sharukhkhanajm** [prisma/prisma#19337] [Q&A]
Hi,
here are my models

```prisma
model Tag {
  id        String   @id @default(auto()) @map("_id") @db.ObjectId
  name      String
  type        EntityType
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  // Relationships
  materials   Material[] @relation(fields: [materialIds], references: [id])
  materialIds String[]   @db.ObjectId

  // Indexes and Uniques constraints
  @@unique([name, type])
}

model Material {
  id        String   @id @default(auto()) @map("_id") @db.ObjectId
  name      String
  quantity  Float
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  // Relationships  
  tags   Tag[]    @relation(fields: [tagIds], references: [id])
  tagIds String[] @db.ObjectId
}

```

In the current model structure the `Material` can have multiple `Tag` in it.
Now when I create a new `Material` I connect the `Tag` with the `Material`.

The Problem I am facing is:

So when I delete the `Material` the `Tag` model is still connected with the deleted `Material`.
How can I remove this connection when deleting the `Material`?

<img width="1433" alt="image" src="https://github.com/prisma/prisma/assets/47321682/9dfbb552-5927-4eb7-9d42-97b26393b21d">

Hope I am able to convey my message.

Thanks.

**Comment by sharukhkhanajm** [comment]
Hi all,

I don't know if this solution is correct, but it's working until I get a better solution.

Here is how I tried to achieve it.

When deleting the `Material`

First I found all the `Tag` associated with the material

```js
const materialTags = await prisma.tag.findMany({
    where: {
      materials: {
        some: {
          id,
        },
      },
    },
  });

```

Then I used transaction and first updated the `Material` and remove the `Tag` connection and deleted it.

```js
const [, material] = await prisma.$transaction([
    prisma.material.update({
      where: {
        id,
      },
      data: {
        tags: {
          disconnect: materialTags.map((materialTag) => ({
            id: materialTag.id,
          })),
        },
      },
    }),
    prisma.material.delete({
      where: {
        id,
      },
    }),
  ]);

```


The same thing I did when deleting the `Tag`

```js
const tagMaterials = await prisma.material.findMany({
      where: {
        tags: {
          some: {
            id,
          },
        },
      },
    })

```

```js
const [, tag] = await prisma.$transaction([
    prisma.tag.update({
      where: {
        id,
      },
      data: {
        materials: {
          disconnect: tagMaterials.map((material) => ({
            id: material.id,
          })),
        },
      },
    }),
    prisma.tag.delete({
      where: {
        id,
      },
    }),
  ]);

```

I know it's tedious to write but that's the only solution I could come up with.

Hope to get a better implementation.

Thanks.

**Comment by sharukhkhanajm** [comment]
**Just an update**:

This issue does not seem to happen in **Postgresql**  

I think this issue only exists in **MongoDB** as in NoSQL there is no foreign key constraint.

It's just my guess rectify me if I am wrong :)

**Reply by ludralph** [reply]
Disconnect records applies to all databases. You can take a look at this section of the [documentation](https://www.prisma.io/docs/orm/prisma-client/queries/relation-queries#disconnect-all-related-records)

**Comment by ludralph** [comment]
Hi @sharukhkhanajm 👋 

Your approach to disconnecting related records before deletion is consistent. You are using transactions to ensure that both the disconnection of related records and the deletion of the main record are executed as part of a single atomic operation, which is a recommended practice.

 If you want to remove all materials from a tag, you can use this:
```js
const data = await prisma.tags.update({
  where: {
    id,
  },
  data: {
    materials: {
      set: [],
    },
  },
});
```
You can also do the same thing to disconnect all tags from a material. See the [documenattion](https://www.prisma.io/docs/orm/prisma-client/queries/relation-queries#disconnect-all-related-records) to learn more.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by sharukhkhanajm** [comment]
Hi @ludralph Thanks for the reply.
Closing this ticket.
